### ISSUE TITLE
[combobox] Allow Combobox.Clear to be focusable and keyboard navigatable

### ISSUE THREAD

**Issue by StefanDejanovskiOcuco** [mui/base-ui#3630]
# Feature request

Allow the "Clear selection" button to be accessible by keyboard navigation (tab and arrow keys)

## Summary

Currently `Combobox.Clear` renders with `tabIndex={-1}` which makes it inaccessible, it would be great if the component has a prop that allows the users to access it via keyboard

## Examples in other libraries

https://www.mongodb.design/component/combobox/live-example

## Motivation

Currently, keyboard-only users cannot clear their selections without using a mouse, creating a significant accessibility barrier.

Current workaround limitations:
- Users must either use a mouse (not possible for some users)
- Delete selections one by one (time-consuming for multiple selections)
- Navigate away and return to reset the field (disrupts workflow)

**Comment by mnajdova** [comment]
The cleara button is a mouse equivalent of the <kbd>Del</kbd> keyboard key, so there is no need to be focusable inside it will create more friction as each selected item will need to either act as a trap focus zone or have two tab stops. 

So, in practice, Arrow left/right allow keyboard user to focus each combobox item that is selected, and Del removed the item.

**Comment by StefanDejanovskiOcuco** [comment]
I agree that two focusable elements will create friction for the `Combobox.Chip` component,

but this request is for the  `Combobox.Clear` component not `Combobox.ChipRemove`

**Comment by atomiks** [comment]
`<Combobox.Clear tabIndex={0}>` works, no? But yes, the problem here is multiple tab stops in a form. Ideally each field is one single tab stop. For keyboard users (when not using a screen reader), <kbd>Esc</kbd> will clear the value in the input (when the popup is closed), or deleting all the text in the input

**Comment by sarthakmalik0810** [comment]
@atomiks Yes <kbd>Esc</kbd> works, but when you are in a `Dialog`, <kbd>Esc</kbd> would actually close the dialog as well. Atleast thats what happened when I used it in a Radix `Dialog`.

**Comment by StefanDejanovskiOcuco** [comment]
@atomiks `<Combobox.Clear tabIndex={0}>` doesn't work fully, this makes the element focusable via `Tab`, but the keyboard navigation with left and right arrow keys is done by the library and won't focus the `Clear` component.
`Esc` is great, but `Clear` is an interactive component and it should be able to be navigated to via keyboard consistently, as it is with the `Chips` elements.

@sarthakmalik0810 I think that's expected, the event should be prevented from bubbling up in order for the Dialog to stay open.

**Comment by atomiks** [comment]
It does prevent bubbling on <kbd>Esc</kbd>, and works with our Dialog at least

> keyboard navigation with left and right arrow keys is done by the library and won't focus the Clear component

What is the reason the Clear button should be part of the composite loop? Other examples of this pattern (e.g. Chakra UI and Material UI) don't include the clear button as part of the focus loop or tab order

A blog post https://scottaohara.github.io/clear-text-field-button/ explains why the clear button is not in the tab order:

> The clear button is purposefully not in the tabbing order of the web page. Keyboard users can use the Delete or Backspace keys (or first select all the text field's text and then use these keys) to delete the contents of the text field.
> 
> Important: if the clear button was part of the web page's tabbing order, every text field a user had entered data into would then require that they tab past the clear button that they have no use for. This clear button being useless to them if they had filled out the field with the data they intended. Adding clear buttons to the web page's tabbing order is not recommended for this reason, as it has the potential to require two keyboard focus stops for every text field on the page that has a clear button.



> The HTML <input type="search"> also excludes the clear button from the tab sequence https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/search.

https://github.com/mui/material-ui/issues/44936#issuecomment-2604768121

**Comment by StefanDejanovskiOcuco** [comment]
It's interesting that the input with type="search" works in this way, though this was introduced in 2015 and seems outdated.

IMO, I don't agree with the blog post, and I think this is not great UX. There isn't a way for the screen reader to announce the clear button, and there IS obviously a button which is intentionally hidden from screen readers (i.e., people who rely on screen readers).

Hiding it from screen reader, on `blur` or on `mouseout` just makes it less discoverable.

Also, MDN states this: https://developer.mozilla.org/en-US/docs/Web/Accessibility/Guides/Understanding_WCAG/Keyboard#click%E2%80%A6 (Currently the `Combobox.Clear` is a button → clickable element, and has `tabIndex={-1}` which removes the interactive semantics)

https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html - You can argue that the `Esc` key is enough to satisfy this constraint, but providing multiple paths is better UX.

**Comment by silviuaavram** [comment]
@StefanDejanovskiOcuco the screen readers don't need to announce anything about the clear button. They announce it's a combobox. Screen reader users know that a combobox content could be removed when pressing escape. If you take a look at the [ARIA examples](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/) you won't see any clear button at all. The clear button is just a helper button that has been introduced for mouse users, nothing more. It improves their UX by just clicking the button instead of clicking the input, selecting all content, and pressing delete.

Also, regarding the mongo db example, their UX decision involves a learning curve for users. In their combobox, you could clear the content with Esc, but not after you selected an option, you can only use the X button for that, which only appears after selection. Anyway, it's their UX decision, it's not a standard combobox pattern, which is the one we opt for.

**Comment by StefanDejanovskiOcuco** [comment]
@silviuaavram I think you missed the point, what I'm trying to say is "multiple paths to the clear functionality is more UX friendly then only one path"

1. I'm not saying your implementation is incorrect, this is a feature request not a bugfix.
2. I'm not saying mongodb's implementation is better, its just an example of arrow navigation to the clear button, nothing related to selecting options.
3. Don't rely solely on the link you provided, the first sentence in there states this "The code in this example is not intended for production environments.", so it's a guide not a specification.

Either way its your decision, if you feel so strongly about having only `Esc` access the clear functionality you can close this thread.
